### PR TITLE
Adding lint test for excessive LOC.

### PR DESCRIPTION
**Pull request by avborhanian**
This is a WIP for #2377. Just wanted to pull in because I had a few questions:

1. Is it okay that I'm approaching this via counting by looking at each line in the snippet instead of looking at the AST tree? If there's another way to do it, I want to make sure I'm doing the correct way, but I wasn't sure since the output AST JSON doesn't seem to contain whitespace.

2. My function is definitely going to trigger the lint, so also wanted to see if there was something obvious I could do to reduce it.

3. Are the two tests fine, or is there something obvious I'm missing?

4. Obviously bigger question - am I approaching the line count correctly. Current strategy is count a line if it contains some code, so skip if it's just comments or empty.

### CODE REVIEW

**Comment by oli-obk** [comment]
1. yes
2. I'll comment inline
3. please add them to their own file. We're trying to have many small tests files instead of few large ones
4. makes sense

File: clippy_lints/src/functions.rs [hunk 1]
```diff
@@ -138,6 +155,81 @@ impl<'a, 'tcx> Functions {
...
+        let code_snippet = snippet(cx, span, "..");
+        let mut line_count = 0;
+        let mut in_comment = false;
+        for line in code_snippet.lines() {
+            let mut s_line = line.trim_left();
+            if in_comment {
+                if !s_line.contains("*/") { continue; }
+                in_comment = false;
+                let end_comment_loc = s_line.find("*/").expect("*/ should be present.");
+                s_line = &s_line[end_comment_loc..];
```
Comment (oli-obk) [inline clippy_lints/src/functions.rs]: needs another trim?

File: clippy_lints/src/functions.rs [hunk 2]
```diff
@@ -138,6 +155,81 @@ impl<'a, 'tcx> Functions {
...
     }
 
+    fn check_line_number(self, cx: &LateContext, span: Span) {
+        let code_snippet = snippet(cx, span, "..");
+        let mut line_count = 0;
+        let mut in_comment = false;
+        for line in code_snippet.lines() {
+            let mut s_line = line.trim_left();
+            if in_comment {
+                if !s_line.contains("*/") { continue; }
```
Comment (oli-obk) [inline clippy_lints/src/functions.rs]: You can do

```rust
let end_comment_loc = match s_line.find("*/") {
    Some(pos) => pos,
    None => continue,
};
```

which also saves you the `unwrap` below

File: clippy_lints/src/functions.rs (hunk 1 above)
Comment (oli-obk) [reply clippy_lints/src/functions.rs]: Maybe don't trim at all until just before line 170?

File: clippy_lints/src/functions.rs [hunk 3]
```diff
@@ -138,6 +155,81 @@ impl<'a, 'tcx> Functions {
...
+            if s_line.is_empty() || s_line.starts_with("//") {
+                continue;
+            }
+            if s_line.contains("/*") {
+                let mut count_line: bool = !s_line.starts_with("/*");
+                let mut closing_comments = s_line.match_indices("*/");
+                let mut open_comments = s_line.match_indices("/*");
+                let mut single_comment = s_line.find("//");
+                let mut curr_idx: usize = 0;
+                loop {
```
Comment (oli-obk) [inline clippy_lints/src/functions.rs]: That's a lot of complexity. Maybe just treat lines containing multiple comment opening and closing things as non-comment lines. This is a very degenerate case that is rare in real code.

File: clippy_lints/src/functions.rs [hunk 4]
```diff
@@ -138,6 +155,81 @@ impl<'a, 'tcx> Functions {
...
+                        }
+                    }
+                }
+                if count_line {
+                    line_count += 1;
+                }
+            }
+        }
+
+        if line_count > 50 {
```
Comment (oli-obk) [inline clippy_lints/src/functions.rs]: can you make the value configurable via `clippy.toml`?

File: clippy_lints/src/functions.rs (hunk 3 above)
Comment (avborhanian) [reply clippy_lints/src/functions.rs]: Makes sense. So maybe just do this based on counts of the opens and closes?
multiple opens or multiple closes -> counts as a code line, out of comment.
one open, no close -> start of a multiline comment, so I'll toggle that in_comment var
one open, one close -> just check if remaining is empty or //, and increment line count if not so.

File: clippy_lints/src/functions.rs [hunk 5]
```diff
@@ -28,6 +28,22 @@ declare_clippy_lint! {
...
+/// due to having to look at a larger amount of code to understand what the
+/// function is doing. Consider splitting the body of the function into
+/// multiple functions.
+///
+/// **Known problems:** None.
+///
+/// ```
+declare_clippy_lint! {
+    pub TOO_MANY_LINES,
+    complexity,
```
Comment (oli-obk) [inline clippy_lints/src/functions.rs]: Can you make this a `pedantic` lint? I don't think we can get the community to agree on a threshold below 200 ;)

**Comment by oli-obk** [comment]
Let's retrigger CI

**Comment by mati865** [comment]
Triggered itself in `matches.rs` test, you will have to allow it there:
```
+error: This function has a large number of lines.
+   --> $DIR/matches.rs:126:1
+    |
+126 | / fn match_wild_err_arm() {
+127 | |     let x: Result<i32, &str> = Ok(3);
+128 | |
+129 | |     match x {
+...   |
+207 | |     }
+208 | | }
+    | |_^
+    |
+    = note: `-D too-many-lines` implied by `-D warnings`
```

----
Your test doesn't work, maybe you have to add two prints so function body exceeds 51?
```
diff of stderr:
-error: This function has a large number of lines.
-   --> $DIR/functions_maxlines.rs:57:1
-    |
-57  | / fn bad_lines() {
-58  | |     println!("This is bad.");
-59  | |     println!("This is bad.");
-60  | |     println!("This is bad.");
-...   |
-107 | |     println!("This is bad.");
-108 | | }
-    | |_^
-    |
-    = note: `-D too-many-lines` implied by `-D warnings`
-
-error: aborting due to previous error
```

**Comment by oli-obk** [comment]
Dogfood tests die on this :D Clippy itself violates that rule so many times... https://travis-ci.org/rust-lang-nursery/rust-clippy/jobs/397138977#L1006

Not sure if we should increase the limit, allow the lint inside clippy_lints, make it a restriction lint, or any combination of the above.

**Comment by mati865** [comment]
@oli-obk I just noticed restriction group is not mentioned in `README.md`, guess it should be there.

**Comment by avborhanian** [comment]
So for those tests, the line counts were on average 101 lines. Median was 68 lines. Mode was 71 lines. If I increased the limit to trigger on line counts greater than or equal to 101 lines, it dropped the number of failures from 72 to 17, in 14 different files instead of 45 different files.

I think it would make sense to make it allowed inside clippy_lints. I'm not sure what the restriction limit does though, not as familiar with that.

**Comment by oli-obk** [comment]
The restriction group means it's a lint that you have to enable explicitly.

Even with the guidelines from the clippy RFC I'm not sure in which group to place it though ;) It has no false positives, but it can be inactionable (imagine a function exhaustively matching on a 50 variant enum with a few lines of actions per variant).

The 17 functions that you found, did they look like they should be split up?

**Comment by fabric-and-ink** [comment]
Is it possible to deactivate a lint for a specific block? Then we could silence the lint for those 17 functions and postpone a refactoring.

**Comment by oli-obk** [comment]
Sure, just allow it. Or just make it a restriction lint so it needs explicit `warn` attributes to get turned on.

**Comment by phansch** [comment]
@avborhanian would you like to finish up this PR? It would need a rebase and the 17 functions would have to be annotated with `#[allow(clippy::too_many_lines)]` to get started.

**Comment by flip1995** [comment]
Ping from triage @avborhanian: It looks like this PR hasn't received any updates in a while, so I'm closing it per [our new guidelines](https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md#issue-and-pr-triage). Thank you for your contributions and please feel free to re-open in the future.

**Comment by flip1995** [comment]
Thanks for the update @avborhanian! It seems that the rebase somehow failed. Could you try to fix this? After that we're happy to make another review.

**Comment by avborhanian** [comment]
Just curious, is there a way to adjust a configuration variable at the individual test level? Feel like it could be useful for either decreasing the lines I'm testing/making sure different values work as expected.

**Comment by flip1995** [comment]
Yes there is! You can look at `tests/ui-toml` for examples.

**Comment by avborhanian** [comment]
Looks like the current version of nightly is encountering an ICE with the regex crate, but I'm sure I'm still messing up formatting.

File: tests/ui/functions_maxlines.rs [hunk 6]
```diff
@@ -1,162 +0,0 @@
-#![warn(clippy::all, clippy::pedantic)]
```
Comment (flip1995) [inline tests/ui/functions_maxlines.rs]: Oh you could have left this test in, so we know that it also works with the default configuration.

Comment (avborhanian) [reply tests/ui/functions_maxlines.rs]: Fair! I can add it back. I was just thinking about keeping everything in one place, but this makes sense.

**Comment by phansch** [comment]
It looks like it currently causes an ICE with the `bitflags` crate. A minimal repro:

```rust
#[macro_use]
extern crate bitflags;

#[allow(non_upper_case_globals)]
#[allow(unused)]
mod module {
    bitflags! {
        pub struct A: u16 {
            const Abc     = 1 << 0;
        }
    }
}
```

Adding that to the UI test and an `bitflags = "1.0"` to the dev-dependencies in `Cargo.toml` should make it reproducible for you.

I think returning when `in_external_macro` returns true should fix the crash.

**Comment by avborhanian** [comment]
I feel silly for asking this but is the top level Cargo.toml not the right one to be adding the bitflags dependency to? I'm hitting an error when trying to run the test locally.

```
---- [ui] ui\functions_maxlines.rs stdout ----
normalized stderr:
error[E0464]: multiple matching crates for `bitflags`
  --> $DIR/functions_maxlines.rs:4:1
   |
LL | extern crate bitflags;
   | ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: candidates:
           crate `bitflags`: /?/.../rust-clippy/target/debug/deps/libbitflags-6b531e3975302896.rlib
           crate `bitflags`: /?/.../.rustup/toolchains/nightly-x86_64-pc-windows-msvc/lib/rustlib/x86_64-pc-windows-msvc/lib/libbitflags-ecbe27eb754e0c5e.rlib

error[E0463]: can't find crate for `bitflags`
  --> $DIR/functions_maxlines.rs:4:1
   |
LL | extern crate bitflags;
   | ^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to 2 previous errors-
```

**Comment by avborhanian** [comment]
@phansch Couldn't figure out a way to get around the issue, so I ended up removing it for now. Hope that's okay. Updated my branch to have the new changes.

**Comment by phansch** [comment]
@avborhanian thanks! it just needs a last `cargo fmt` in the tests: https://travis-ci.com/rust-lang/rust-clippy/jobs/174590331#L1038-L1062

**Comment by avborhanian** [comment]
I see the lines now. I'm going to add #[rustfmt::skip] to those if that's cool, because the original reason for those tests was to see the weird cases with comments, as unrealistic as they were (and obviously incorrect for rustfmt). If not, I can just fix the format and we can still use those tests as is anyways, as extra line count checks.

File: clippy_lints/src/types.rs [hunk 7]
```diff
@@ -240,8 +240,9 @@ fn match_type_parameter(cx: &LateContext<'_, '_>, qpath: &QPath, path: &[&str])
 ///
 /// The parameter `is_local` distinguishes the context of the type; types from
 /// local bindings should only be checked for the `BORROWED_BOX` lint.
-fn check_ty(cx: &LateContext<'_, '_>, hir_ty: &hir::Ty, is_local: bool) {
-    if in_macro(hir_ty.span) {
+#[allow(clippy::too_many_lines)]
+fn check_ty(cx: &LateContext<'_, '_>, ast_ty: &hir::Ty, is_local: bool) {
```
Comment (phansch) [inline clippy_lints/src/types.rs]:
```suggestion
fn check_ty(cx: &LateContext<'_, '_>, hir_ty: &hir::Ty, is_local: bool) {
```

File: clippy_lints/src/types.rs [hunk 8]
```diff
@@ -240,8 +240,9 @@ fn match_type_parameter(cx: &LateContext<'_, '_>, qpath: &QPath, path: &[&str])
...
+    if in_macro(ast_ty.span) {
```
Comment (phansch) [inline clippy_lints/src/types.rs]:
```suggestion
    if in_macro(hir_ty.span) {
```

**Comment by phansch** [comment]
Thanks for your patience! Everything LGTM now.

@bors r+

**Comment by avborhanian** [comment]
I was about to say the same thing - 7 month long pull request, thanks for the patience. :D

**Pull request merged**